### PR TITLE
Factor out more Travis code and update GitHub Actions 

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Makes more CI jobs compatible with GitHub Actions.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: ci/travis/install-toolchains.sh
Comment (simon-mo) [inline ci/travis/install-toolchains.sh:37]: if we don't do this, would this invalidate Bazel's cache every time? or this is just a one time cost

File: ci/travis/ci.sh
Comment (simon-mo) [inline ci/travis/ci.sh:150]: ? what's this section about

File: ci/travis/ci.sh
Comment (simon-mo) [inline ci/travis/ci.sh:67]: what does this section do?

File: ci/travis/ci.sh
Comment (simon-mo) [inline ci/travis/ci.sh:72]: i'm pretty sure github action has some feature like this? fail fast or something?

Comment (mehrdadn) [reply ci/travis/ci.sh:72]: I don't think it does? And I think you mean `continue-on-error`? Fail-fast means that if a job fails then the entire workflow fails, but here we want it to succeed. `continue-on-error` would make it succeed but errors can occur due to lots of other issues, which we wouldn't want to ignore.

File: ci/travis/ci.sh
Comment (mehrdadn) [reply ci/travis/ci.sh:150]: It's where I'm adding code to run tests, build wheel, etc. I have that coming in the next PR.

Comment (simon-mo) [reply ci/travis/ci.sh:150]: - Then let's add it in the next PR, make this one cleaner. 
- asking again, any chance we can move these to python?

File: ci/travis/ci.sh
Comment (mehrdadn) [reply ci/travis/ci.sh:67]: It emits the assignments (`RAY_CI_XYZ=1`) into the `.bashrc`. I could keep them as exports, but I'm trying to architecture it so that we don't try to refer to them from external scripts later, since those variables are CI-specific and shouldn't be depended on by scripts that could conceivably be called without these variables being defined. Let's make sure the tests are working before we merge this though!

File: ci/travis/ci.sh
Comment (mehrdadn) [reply ci/travis/ci.sh:150]: - Sure

- I can do that once we've factored out the code from Travis, but rewriting in Python seems orthogonal to being able to call it from GitHub Actions? And if I try to rewrite everything in Python at the same time as I'm moving the scripts around, it's very likely something is going to break. We can revisit that idea once the reshuffling is done?

File: ci/travis/install-toolchains.sh
Comment (mehrdadn) [reply ci/travis/install-toolchains.sh:37]: It's a fixed cost, but might happen more than once per build, depending on how many different PATHs we call Bazel with (for testing, wheels, etc.)

That's secondary to the concern of not having to reload `PATH` everywhere though. It's so much easier if we don't even have to worry about managing `PATH` everywhere.

File: python/setup.py
Comment (mehrdadn) [reply python/setup.py:18]: I think you're thinking of `.pyx` and `.pxi` and such? [`.pyd` is the extension Python requires Windows for shared libraries.](https://docs.python.org/3/faq/windows.html#is-a-pyd-file-the-same-as-a-dll)

File: ci/travis/ci.sh
Comment (simon-mo) [reply ci/travis/ci.sh:72]: If `continue-on-error` is set to false, all the following steps will be skipped automatically right?

File: ci/travis/ci.sh
Comment (simon-mo) [inline ci/travis/ci.sh:148]: why do we need this extra _main?

File: ci/travis/install-toolchains.sh
Comment (simon-mo) [reply ci/travis/install-toolchains.sh:37]: can't we just `export PATH=...` in .bashrc?

File: python/setup.py
Comment (simon-mo) [reply python/setup.py:18]: let's put this link as a comment in this file so future maintainers know what's going on

File: ci/travis/install-toolchains.sh
Comment (mehrdadn) [reply ci/travis/install-toolchains.sh:37]: I just thought it'd be better to not modify `.bashrc` and `PATH` everywhere unless I really need to. Note that this is Windows-only. But I can look into doing that instead.

File: ci/travis/ci.sh
Comment (mehrdadn) [reply ci/travis/ci.sh:72]: I thought it only skips continues this step if there's an error and then it moves onto the next one. But regardless, I'm saying the bigger issue is something else: using it would mask _all_ errors that step, which we really don't want. We only want to exit the build job under a very particular set of conditions, which is that some flags determine that the job should be skipped. This seemed like the easiest way of achieving that without having to put a lot of boilerplate in every CI's configuration files.

File: python/setup.py
Comment (mehrdadn) [reply python/setup.py:18]: Okay sure, I'll add one.

File: ci/travis/ci.sh
Comment (mehrdadn) [reply ci/travis/ci.sh:148]: I'll need it in the next PR since it where I'll call `reload_env`, but I can take it out for now.

BTW I replied to your other comments, but somehow they're not showing up under your review for me. You can click on the other threads to see them.

**Review by simon-mo — APPROVED** [review]

**Pull request merged**
